src/swupdd-main.c

```diff
@@ -22,751 +18,765 @@
-        errmsg[0] = error->message;
-        clr_software_update__emit_request_completed (object,
-                                                     dstate->method,
-                                                     -1,
-                                                     (const gchar * const *) errmsg);
-        g_error_free (error);
-    }
-    else
-    {
-        g_subprocess_communicate_utf8_async (dstate->process,
-                                             NULL,
-                                             NULL,
-                                             _on_process_done,
-                                             dstate);
-    }
-
-    g_free (argv);
-    g_list_free_full (args, g_free);
-
-    return TRUE;
+	daemon_state_t *context = userdata;
+	int r = 0;
+	char buffer[PIPE_BUF + 1];
+	ssize_t count;
+	count = read(fd, buffer, PIPE_BUF);
+	if (count > 0) {
+		fwrite(buffer, 1, count, stdout);
+		fflush(stdout);
+		buffer[count] = '\0';
+		r = sd_bus_emit_signal(context->bus,
+				       "/org/O1/swupdd/Client",
+				       "org.O1.swupdd.Client",
+				       "childOutputReceived", "s", buffer);
+		if (r < 0) {
+			ERR("Failed to emit signal: %s", strerror(-r));
+		}
+	} else if (count < 0) {
+		r = -errno;
+		ERR("Failed to read pipe: %s", strerror(errno));
+	} else {
+		close(fd);
+		sd_event_source_unref(s);
+	}
+
+	return r < 0 ? EXIT_FAILURE : EXIT_SUCCESS;
 }
 
-
-static gboolean _on_cancel (ClrSoftwareUpdate *object,
-                            GDBusMethodInvocation *invocation,
-                            gboolean arg_force,
-                            gpointer user_data)
+static int on_child_exit(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata)
 {
-    daemon_state_t *dstate = (daemon_state_t *) user_data;
-
-    g_message ("cancel");
-    clr_software_update__complete_cancel (object, invocation,
-                                          (dstate->process != NULL));
-    if (!dstate->process)
-        return FALSE;
-
-    if (arg_force)
-        g_subprocess_force_exit (dstate->process);
-    else
-        g_subprocess_send_signal (dstate->process, SIGINT);
-
-    clr_software_update__emit_request_completed (object,
-                                                 "cancel",
-                                                 0,
-                                                 _empty_strv);
-    return TRUE;
+	daemon_state_t *context = userdata;
+	int child_exit_status;
+	int status = 0;
+	int r;
+
+	method_t child_method = context->method;
+	assert(child_method);
+
+	if (si->ssi_code == CLD_EXITED) {
+		status = si->ssi_status;
+	} else {
+		DEBUG("Child process was killed");
+		/* si->ssi_status is signal code in this case */
+		status = 128 + si->ssi_status;
+	}
+
+	/* Reap the zomby */
+	assert(si->ssi_pid == context->child);
+	pid_t ws = waitpid(si->ssi_pid, &child_exit_status, 0);
+	assert(ws != -1);
+
+	context->child = 0;
+	context->method = METHOD_NOTSET;
```

Comment:
It's closed in `on_childs_output()` upon receiving `EOF` which happens before this handler is called.